index.rst

```diff
@@ -48,9 +48,118 @@
+
+We most definitely do not want to maintain a separate container build
+per user.  Because of the way we are currently handling file permissions
+and collaboration within the RSP, we create a Unix user with UIDs and
+GIDs mapped from our authentication provider, and then we provision a
+home directory for that user if it does not exist.
+
+Currently, we use the ``provisionator`` (``769:769``) account to perform
+those actions, which are privileged.  In order to do that, the container
+build writes sudoers rules to allow the provisionator to act.  It also
+allows the provisionator to write a rule (via a script) that allows the
+provisionator to become the target user before starting Jupyterlab.
+
+This requires ``sudo`` in order to work, and therefore the Lab
+containers cannot be run (at least in environments that require
+filesystem provisioning) with ``allowPrivilegeEscalation: False``.  This
+violates the third item of section 4.1 of `SQR-048`_.
+
+.. _SQR-048: https://sqr-048.lsst.io/
+
+The Partial Solution
+====================
+
+A year or more ago we introduced a ``NO_SUDO`` flag environment variable
+that can be passed into the container, which bypasses all of the sudo
+steps to escalate privilege.
+
+This comes with one major caveat: it only works if there is a
+pre-provisioned user home directory.  This is fine at NCSA (which is
+where the rule was intended for use) since user home space is
+provisioned as part of account creation and NCSA forces all RSP users to
+have NCSA accounts; it is not fine in a GKE environment using GitHub as
+its authentication back-end.  We have recently proven that everything
+still works as expected (assuming a pre-provisioned filesystem) if the
+container is run with ``allowPrivilegeEscalation: False`` in its
+Kubernetes ``securityContext``.
+
+Recently, in order to allow loading the Multus CNI into containers in
+Rubin-managed Kubernetes environments, necessary for control of
+instruments on multicast networks with SAL, Josh Hoblitt introduced a
+container that must be run as a privileged initContainer inside the
+user's Lab pod before starting the Lab.  This change is already present
+in the mainline Nublado environment.
+
+A Fuller Solution
+=================
+
+Since we've worked out how to run privileged initContainers, all we
+would need to do is to move the provisioning steps into a second
+initContainer, and invoke it (or not) at user pod creation time.  We
+could either restrict its privilege still further by doing the same set
+of semi-privileged user plus sudoers rules we currently do, or just let
+the initContainer run as root, which would simplify the logic within
+considerably.
+
+Then, in all cases, we could invoke the Lab pod directly with the
+correct userid and set of groups, having provisioned those in ConfigMaps
+to mount into ``/etc/passwd`` and ``/etc/group`` (``NO_SUDO`` already
+does this), with the only piece that still requires privilege being the
+creation of user filesystems.
```

Comment:
For future documentation purposes, it would be nice to add some more detail of how this works. I assume that the Hub, when spawning the pod, creates a new `ConfigMap` with four keys, one each for `passwd`, `shadow`, `groups`, and `gshadow`, whose values are a fixed template taken from some recent build of the lab container plus a line for the user and lines for each of that user's groups.

Presumably at some point we could replace the `groups` and `gshadow` entries with some fixed `ConfigMap` that contains all of the groups the system is aware of, once we have a group system that can easily return such a thing. That would be more user-friendly since it would allow resolution of GIDs to names for groups the user is not a member of (helpful for the user to know which groups to request access to).

Reply:
Your assumption is correct, and, yes, we could do that, except that I am generally a fan of individual per-user groups as the user's default group.  But...sure, we could write a ConfigMap that is the union of all the groups the system knows about and the new individual-user group.